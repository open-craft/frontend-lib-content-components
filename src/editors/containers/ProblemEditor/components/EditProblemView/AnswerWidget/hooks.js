import { useState, useEffect } from 'react';
import { StrictDict } from '../../../../../utils';
import * as module from './hooks';
import { actions } from '../../../../../data/redux';

export const state = StrictDict({
  isFeedbackVisible: (val) => useState(val),
});

export const removeAnswer = ({ answer, dispatch }) => () => {
  dispatch(actions.problem.deleteAnswer({ id: answer.id }));
};

export const setAnswer = ({ answer, hasSingleAnswer, dispatch }) => (payload) => {
  dispatch(actions.problem.updateAnswer({ id: answer.id, hasSingleAnswer, ...payload }));
};

export const prepareFeedback = (answer) => {
  const [isFeedbackVisible, setIsFeedbackVisible] = module.state.isFeedbackVisible(false);
  useEffect(() => {
    // Show feedback fields if feedback is present
    const isVisible = !!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback;
    setIsFeedbackVisible(isVisible);
  }, [answer]);

  const toggleFeedback = (open) => {
    // Do not allow to hide if feedback is added
    if (!!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback) {
      setIsFeedbackVisible(true);
      return;
    }
    setIsFeedbackVisible(open);
  };
  return {
    isFeedbackVisible,
    toggleFeedback,
  };
};

export default {
  state, removeAnswer, setAnswer, prepareFeedback,
};
