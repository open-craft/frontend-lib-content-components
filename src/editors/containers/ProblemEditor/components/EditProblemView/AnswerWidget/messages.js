export const messages = {
  answerWidgetTitle: {
    id: 'authoring.answerwidget.answer.answerWidgetTitle',
    defaultMessage: 'Answers',
    description: 'Main title for Answers widget',
  },
  addAnswerButtonText: {
    id: 'authoring.answerwidget.answer.addAnswerButton',
    defaultMessage: 'Add answer',
    description: 'Button text to add answer',
  },
  answerTextboxPlaceholder: {
    id: 'authoring.answerwidget.answer.placeholder',
    defaultMessage: 'Enter an answer',
    description: 'Placeholder text for answer option text',
  },
  feedbackPlaceholder: {
    id: 'authoring.answerwidget.feedback.placeholder',
    defaultMessage: 'Feedback message',
    description: 'Placeholder text for feedback text',
  },
  feedbackToggleIconAltText: {
    id: 'authoring.answerwidget.feedback.icon.alt',
    defaultMessage: 'Toggle feedback',
    description: 'Alt text for feedback toggle icon',
  },
  answerDeleteIconAltText: {
    id: 'authoring.answerwidget.answer.delete.icon.alt',
    defaultMessage: 'Delete answer',
    description: 'Alt text for delete icon',
  },
  selectedFeedbackLabel: {
    id: 'authoring.answerwidget.feedback.selected.label',
    defaultMessage: 'Show following feedback when {answerId} {boldunderline}:',
    description: 'Label text for feedback if option is selected',
  },
  selectedFeedbackLabelBoldUnderlineText: {
    id: 'authoring.answerwidget.feedback.selected.label.boldunderline',
    defaultMessage: 'is selected',
    description: 'Bold & underlined text for feedback if option is selected',
  },
  unSelectedFeedbackLabel: {
    id: 'authoring.answerwidget.feedback.unselected.label',
    defaultMessage: 'Show following feedback when {answerId} {boldunderline}:',
    description: 'Label text for feedback if option is not selected',
  },
  unSelectedFeedbackLabelBoldUnderlineText: {
    id: 'authoring.answerwidget.feedback.unselected.label.boldunderline',
    defaultMessage: 'is not selected',
    description: 'Bold & underlined text for feedback if option is not selected',
  },
};

export default messages;
