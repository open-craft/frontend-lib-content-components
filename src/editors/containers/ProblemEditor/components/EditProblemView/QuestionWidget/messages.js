export const messages = {
    questionWidgetTitle : {
        id: 'authoring.questionwidget.question.questionWidgetTitle',
        defaultMessage: 'Question',
        description: 'Question Title',
      },
}
