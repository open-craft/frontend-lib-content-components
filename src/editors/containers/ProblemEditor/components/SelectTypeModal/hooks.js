import {
  useState,
} from 'react';
import { StrictDict } from '../../../../utils';

export const state = StrictDict({
  selected: (val) => useState(val),
});

export default { state };
