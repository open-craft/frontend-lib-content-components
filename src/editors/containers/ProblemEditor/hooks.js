import { useRef, useCallback, useState, useEffect } from 'react';
import { StrictDict } from '../../utils';
import * as module from './hooks';

export const state = StrictDict({
    refReady: (val) => useState(val),
  });

export const problemEditorConfig = ({
    setEditorRef,
    editorRef,
    question,
    updateQuestion
  }) => ({
    onInit: (evt, editor) => {
      setEditorRef(editor);
    },
    initialValue: question ? question : '',
    onFocusOut: () =>{
      let content = editorRef.current.getContent({format: 'text'});
      updateQuestion(content);
    },
  });

  export const prepareEditorRef = () => {
    const editorRef = useRef(null);
    const setEditorRef = useCallback((ref) => {
      editorRef.current = ref;
    }, []);
    const [refReady, setRefReady] = module.state.refReady(false);
    useEffect(() => setRefReady(true), []);
    return { editorRef, refReady, setEditorRef };
  };
