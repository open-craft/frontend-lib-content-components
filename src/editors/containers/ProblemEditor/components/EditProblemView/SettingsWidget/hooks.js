import { useState, useEffect } from 'react';

import * as module from './hooks';
import messages from './messages';
import { actions } from '../../../../../data/redux';
import { isEmpty } from 'lodash-es';

export const state = {
    showAdvanced: (val) => useState(val),
    cardCollapsed: (val) => useState(val),
    summary: (val) => useState(val),
};

export const showAdvancedSettingsCards = () => {
    const [isAdvancedCardsVisible, setIsAdvancedCardsVisible] = module.state.showAdvanced(false);
    return {
        isAdvancedCardsVisible,
        showAdvancedCards: () => setIsAdvancedCardsVisible(true),
    };
}

export const showFullCard = () => {
    const [isCardCollapsed, setIsCardCollapsed] = module.state.cardCollapsed(false);
    return {
        isCardCollapsed,
        toggleCardCollapse: () => setIsCardCollapsed(!isCardCollapsed),
    };
}

export const hintsCardHooks = (hints, dispatch) => {
    const [summary, setSummary] = module.state.summary({ message: "", values: {} })

    useEffect(() => {
        const hintsNumber = hints.length;
        if (hintsNumber == 0) {
            setSummary({ message: messages.noHintSummary, values: {} });
        } else {
            setSummary({ message: messages.hintSummary, values: { hint: hints[0].value, count: (hintsNumber - 1) } });
        }
    }, [hints]);

    const handleAdd = () => {
        let newId = Math.max(...hints.map(hint => hint.id)) + 1
        const hint = { id: newId, value: "" }
        hints = [...hints, hint]
        dispatch(actions.problem.updateSettings({ hints }));
    }


    return {
        summary,
        handleAdd,
    }
}

export const hintsRowHooks = (id, hints, dispatch) => {

    const handleChange = (event) => {
        const value = event.target.value
        hints = hints.map(hint => {
            if (hint.id === id) {
                return { ...hint, value };
            }
            return hint;
        });
        dispatch(actions.problem.updateSettings({ hints }));
    }

    const handleDelete = () => {
        hints = hints.filter((hint) => (hint.id != id));
        dispatch(actions.problem.updateSettings({ hints }));
    }

    return {
        handleChange,
        handleDelete,
    }
}

export const matlabCardHooks = (matLabApiKey, dispatch) => {
    const [summary, setSummary] = module.state.summary({ message: "", values: {}, intl: false })

    useEffect(() => {
        if (isEmpty(matLabApiKey)) {
            setSummary({ message: messages.matlabNoKeySummary, values: {}, intl: true });
        } else {
            setSummary({ message: matLabApiKey, values: {}, intl: false });
        }
    }, [matLabApiKey])

    const handleChange = (event) => {
        dispatch(actions.problem.updateSettings({ matLabApiKey: event.target.value }));
    }

    return {
        summary,
        handleChange,
    }
}

export const randomizationCardHooks = (dispatch) => ({
    handleChange: (event) => {
        dispatch(actions.problem.updateSettings({ randomization: event.target.value }));
    }
})

export const resetCardHooks = (dispatch) => {

    const setReset = (value) => {
        dispatch(actions.problem.updateSettings({ showResetButton: value }));
    }

    return {
        setResetTrue: () => setReset(true),
        setResetFalse: () => setReset(false),
    }
}

export const scoringCardHooks = (scoring, dispatch) => {

    const handleMaxAttemptChange = (event) => {
        let unlimitedAttempts = true;
        const attemptNumber = parseInt(event.target.value);
        if (attemptNumber > 0) {
            unlimitedAttempts = false;
        }
        dispatch(actions.problem.updateSettings({ scoring: { ...scoring, attempts: { number: attemptNumber, unlimited: unlimitedAttempts } } }));
    }

    const handleWeightChange = (event) => {
        dispatch(actions.problem.updateSettings({ scoring: { ...scoring, weight: parseFloat(event.target.value) } }));
    }

    return {
        handleMaxAttemptChange,
        handleWeightChange,
    }
}

export const showAnswerCardHooks = (showAnswer, dispatch) => {

    const handleShowAnswerChange = (event) => {
        dispatch(actions.problem.updateSettings({ showAnswer: { ...showAnswer, on: event.target.value } }));
    }

    const handleAttemptsChange = (event) => {
        dispatch(actions.problem.updateSettings({ showAnswer: { ...showAnswer, afterAttempts: parseInt(event.target.value) } }));
    }

    return {
        handleShowAnswerChange,
        handleAttemptsChange,
    }
}

export const timerCardHooks = (dispatch) => ({

    handleChange: (event) => {
        dispatch(actions.problem.updateSettings({ timeBetween: event.target.value }));
    }
})

export const typeRowHooks = (typeKey, dispatch) => {

    const onClick = () => {
        dispatch(actions.problem.updateField({ problemType: typeKey }));
    }

    return {
        onClick,
    }

}