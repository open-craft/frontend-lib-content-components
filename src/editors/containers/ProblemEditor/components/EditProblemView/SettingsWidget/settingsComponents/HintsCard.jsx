import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, FormattedMessage, intlShape } from '@edx/frontend-platform/i18n';
import { Button } from '@edx/paragon';
import { Add } from '@edx/paragon/icons';
import SettingsOption from '../SettingsOption';
import messages from '../messages';
import { hintsCardHooks, hintsRowHooks } from '../hooks';
import HintRow from './HintRow';

export const HintsCard = ({
  hints,
  updateSettings,
  // inject
  intl,
}) => {
  const { summary, handleAdd } = hintsCardHooks(hints, updateSettings);
  return (
    <SettingsOption
      title={intl.formatMessage(messages.hintSettingTitle)}
      summary={intl.formatMessage(summary.message, { ...summary.values })}
    >
      {hints.map((hint) => (
        <HintRow
          key={hint.id}
          id={hint.id}
          value={hint.value}
          {...hintsRowHooks(hint.id, hints, updateSettings)}
        />
      ))}
      <Button
        className="my-3 ml-2"
        iconBefore={Add}
        variant="tertiary"
        onClick={handleAdd}
      >
        <FormattedMessage {...messages.addHintButtonText} />
      </Button>
    </SettingsOption>
  );
};

HintsCard.propTypes = {
  intl: intlShape.isRequired,
  hints: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
  })).isRequired,
  updateSettings: PropTypes.func.isRequired,
};

export default injectIntl(HintsCard);
