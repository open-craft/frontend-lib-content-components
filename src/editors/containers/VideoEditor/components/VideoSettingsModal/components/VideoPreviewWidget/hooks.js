import messages from '../messages';
import { parseYoutubeId } from '../../../../../../data/services/cms/api';

function getVideoType(videoSource) {
  if (parseYoutubeId(videoSource) !== null) {
    return messages.videoTypeYoutube;
  }
  return messages.videoTypeOther;
}

export default {
  getVideoType,
};
