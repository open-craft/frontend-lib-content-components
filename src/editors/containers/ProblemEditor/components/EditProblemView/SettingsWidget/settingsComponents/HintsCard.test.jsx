import React from 'react';
import { shallow } from 'enzyme';
import { formatMessage } from '../../../../../../../testUtils';
import { HintsCard } from './HintsCard';
import { hintsCardHooks, hintsRowHooks } from '../hooks';
import messages from '../messages';
import { useDispatch } from 'react-redux';

jest.mock('../hooks', () => ({
    hintsCardHooks: jest.fn(),
    hintsRowHooks: jest.fn(),
}));

describe('HintsCard', () => {
    const hint1 = { id: 1, value: "hint1" }
    const hint2 = { id: 2, value: "" }
    const hints_0 = [];
    const hints_1 = [hint1];
    const hints_2 = [hint1, hint2];
    const props = {
        intl: { formatMessage },
        hints: hints_0,
        updateSettings: jest.fn(),
    };

    const hintsRowHooksProps = {
        handleChange: jest.fn().mockName('hintsRowHooks.handleChange'),
        handleDelete: jest.fn().mockName('hintsRowHooks.handleDelete'),
    }
    hintsRowHooks.mockReturnValue(hintsRowHooksProps);

    describe('behavior', () => {
        it(' calls hintsCardHooks when initialized', () => {
            const hintsCardHooksProps = {
                summary: { message: messages.noHintSummary, values: {} },
                handleAdd: jest.fn().mockName('hintsCardHooks.handleAdd'),
            };
            
            hintsCardHooks.mockReturnValue(hintsCardHooksProps);
            shallow(<HintsCard {...props} />);
            const dispatch = useDispatch();
            expect(hintsCardHooks).toHaveBeenCalledWith( hints_0, dispatch );
        });
    });

    describe('snapshot', () => {
        test('snapshot: renders hints setting card no hints', () => {
            const hintsCardHooksProps = {
                summary: { message: messages.noHintSummary, values: {} },
                handleAdd: jest.fn().mockName('hintsCardHooks.handleAdd'),
            };

            hintsCardHooks.mockReturnValue(hintsCardHooksProps);
            expect(shallow(<HintsCard {...props} />)).toMatchSnapshot();
        });
        test('snapshot: renders hints setting card one hint', () => {
            const hintsCardHooksProps = {
                summary: {
                    message: messages.hintSummary,
                    values: { hint: hint1.value, count: 1 }
                },
                handleAdd: jest.fn().mockName('hintsCardHooks.handleAdd'),
            };

            hintsCardHooks.mockReturnValue(hintsCardHooksProps);
            expect(shallow(<HintsCard {...props} hints={hints_1} />)).toMatchSnapshot();
        });
        test('snapshot: renders hints setting card multiple hints', () => {
            const hintsCardHooksProps = {
                summary: {
                    message: messages.hintSummary,
                    values: { hint: hint2.value, count: 2 }
                },
                handleAdd: jest.fn().mockName('hintsCardHooks.handleAdd'),
            };

            hintsCardHooks.mockReturnValue(hintsCardHooksProps);
            expect(shallow(<HintsCard {...props} hints={hints_2} />)).toMatchSnapshot();
        });
    });
});