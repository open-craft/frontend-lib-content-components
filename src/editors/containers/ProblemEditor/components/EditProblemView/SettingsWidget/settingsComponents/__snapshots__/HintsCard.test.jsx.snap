// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HintsCard snapshot snapshot: renders hints setting card multiple hints 1`] = `
<SettingsOption
  summary=" {count, plural, =0 {} other {(+# more)}}"
  title="Hints"
>
  <InjectIntl(ShimmedIntlComponent)
    handleChange={[MockFunction hintsRowHooks.handleChange]}
    handleDelete={[MockFunction hintsRowHooks.handleDelete]}
    id={1}
    key="1"
    value="hint1"
  />
  <InjectIntl(ShimmedIntlComponent)
    handleChange={[MockFunction hintsRowHooks.handleChange]}
    handleDelete={[MockFunction hintsRowHooks.handleDelete]}
    id={2}
    key="2"
    value=""
  />
  <Button
    className="my-3 ml-2"
    onClick={[MockFunction hintsCardHooks.handleAdd]}
    variant="tertiary"
  >
    <FormattedMessage
      defaultMessage="Add hint"
      description="Add hint button text"
      id="authoring.problemeditor.settings.hint.addHintButton"
    />
  </Button>
</SettingsOption>
`;

exports[`HintsCard snapshot snapshot: renders hints setting card no hints 1`] = `
<SettingsOption
  summary="No Hints"
  title="Hints"
>
  <Button
    className="my-3 ml-2"
    onClick={[MockFunction hintsCardHooks.handleAdd]}
    variant="tertiary"
  >
    <FormattedMessage
      defaultMessage="Add hint"
      description="Add hint button text"
      id="authoring.problemeditor.settings.hint.addHintButton"
    />
  </Button>
</SettingsOption>
`;

exports[`HintsCard snapshot snapshot: renders hints setting card one hint 1`] = `
<SettingsOption
  summary="hint1 {count, plural, =0 {} other {(+# more)}}"
  title="Hints"
>
  <InjectIntl(ShimmedIntlComponent)
    handleChange={[MockFunction hintsRowHooks.handleChange]}
    handleDelete={[MockFunction hintsRowHooks.handleDelete]}
    id={1}
    key="1"
    value="hint1"
  />
  <Button
    className="my-3 ml-2"
    onClick={[MockFunction hintsCardHooks.handleAdd]}
    variant="tertiary"
  >
    <FormattedMessage
      defaultMessage="Add hint"
      description="Add hint button text"
      id="authoring.problemeditor.settings.hint.addHintButton"
    />
  </Button>
</SettingsOption>
`;
